### PR TITLE
feat: UpDownCounter, Histogram, tmetrics

### PR DESCRIPTION
**Pull request by ricktu-mw**
- support attributes input for add method in Counter class
- build UpDownCounter and Histogram classes
- build unit tests for meter instruments including Counter, UpDownCounter and Histogram classes

### CODE REVIEW

File: test/tmetrics.m [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+            % test names, added value and attributes in Counter
+
+            metername = "foo";
+            countername = "bar";
+
+            p = opentelemetry.sdk.metrics.MeterProvider();
+            mt = p.getMeter(metername);
+            ct = mt.createCounter(countername);
+
+            % verify MATLAB object properties
```
Comment (duncanpo) [inline test/tmetrics.m:92]: These tests on lines 93-96 are unnecessary since they are duplicates of lines 50-54.

File: test/tmetrics.m [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+            verifyEqual(testCase, string(results.resourceMetrics.scopeMetrics.scope.name), metername);
+
+            % fetch datapoint
+            dp = results.resourceMetrics.scopeMetrics.metrics.sum.dataPoints;
+
+            % verify counter value
+            verifyEqual(testCase, dp.asDouble, sum(vals));
+
+            % verify counter attributes
+            verifyEqual(testCase, string(dp.attributes(1).key), dict_keys(1));
```
Comment (duncanpo) [inline test/tmetrics.m:127]: These seems fragile in the sense that it relies on the results keeping the attributes in the same order. Refer to testAttributes in ttrace.m for a way to write tests that does not assume the attribute order.

File: test/tmetrics.m [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+            ct = mt.createUpDownCounter(countername);
+
+            % verify MATLAB object properties
+            verifyEqual(testCase, mt.Name, metername);
+            verifyEqual(testCase, mt.Version, "");
+            verifyEqual(testCase, mt.Schema, "");
+            verifyEqual(testCase, ct.Name, countername);
+
+            % create testing value 
+            val = 10;
```
Comment (duncanpo) [inline test/tmetrics.m:177]: For UpDownCounter, it makes more sense to test negative values rather than positive values.

File: test/tmetrics.m [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+            % test names, added value and attributes in UpDownCounter
+
+            metername = "foo";
+            countername = "bar";
+
+            p = opentelemetry.sdk.metrics.MeterProvider();
+            mt = p.getMeter(metername);
+            ct = mt.createUpDownCounter(countername);
+
+            % verify MATLAB object properties
```
Comment (duncanpo) [inline test/tmetrics.m:212]: These are exact duplicates of 171-174, so can be skipped.

File: test/tmetrics.m [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+            % verify statistics
+            verifyEqual(testCase, dp.min, val);
+            verifyEqual(testCase, dp.max, val);
+            verifyEqual(testCase, dp.sum, val);
+            
+            % verify count in bucket
+            lower = bounds(1);
+            upper = bounds(2);
+            expect_count = sum(val>lower & val<=upper);
+            verifyEqual(testCase, str2double(counts{2}), expect_count);
```
Comment (duncanpo) [inline test/tmetrics.m:290]: Check all buckets. Don't just check one of them.

File: test/tmetrics.m [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+            verifyEqual(testCase, dp.sum, sum(vals));
+            
+            % verify attributes
+            verifyEqual(testCase, string(dp.attributes(1).key), dict_keys(1));
+            verifyEqual(testCase, string(dp.attributes(1).value.stringValue), dict_vals(1));
+            verifyEqual(testCase, string(dp.attributes(2).key), dict_keys(2));
+            verifyEqual(testCase, string(dp.attributes(2).value.stringValue), dict_vals(2));
+            
+            % verify count in bucket
+            for i = 2:(length(counts)-1)
```
Comment (duncanpo) [inline test/tmetrics.m:338]: Should also manually check the first and last bins.

**Review by duncanpo — CHANGES_REQUESTED** [review]
I suggest fixing a few places in the tests. Please fix those and I will accept the change. Let me know if you have any questions about my comments.

**Review by duncanpo — APPROVED** [review]
Changes look good
